exporter/awss3exporter/sumo_marshaler.go

```diff
@@ -0,0 +1,147 @@
+}
+
+func NewSumoMarshaler() SumoMarshaler {
+	return SumoMarshaler{}
+}
+
+func logEntry(buf *bytes.Buffer, format string, a ...interface{}) {
+	buf.WriteString(fmt.Sprintf(format, a...))
+	buf.WriteString("\n")
+}
+
+func attributeValueToString(v pcommon.Value) string {
+	switch v.Type() {
+	case pcommon.ValueTypeStr:
+		return v.Str()
+	case pcommon.ValueTypeBool:
+		return strconv.FormatBool(v.Bool())
+	case pcommon.ValueTypeBytes:
+		return bytesToString(v.Bytes())
+	case pcommon.ValueTypeDouble:
+		return strconv.FormatFloat(v.Double(), 'f', -1, 64)
+	case pcommon.ValueTypeInt:
+		return strconv.FormatInt(v.Int(), 10)
+	case pcommon.ValueTypeSlice:
+		return sliceToString(v.Slice())
+	case pcommon.ValueTypeMap:
+		return mapToString(v.Map())
+	case pcommon.ValueTypeEmpty:
+		return ""
+	default:
+		return fmt.Sprintf("<Unknown OpenTelemetry attribute value type %q>", v.Type())
+	}
+}
+
+func bytesToString(bs pcommon.ByteSlice) string {
+	var b strings.Builder
+	b.WriteByte('[')
+	for i := 0; i < bs.Len(); i++ {
+		if i < bs.Len()-1 {
+			fmt.Fprintf(&b, "%v, ", bs.At(i))
+		} else {
+			b.WriteString(fmt.Sprint(bs.At(i)))
+		}
+	}
+
+	b.WriteByte(']')
+	return b.String()
+}
+
+func sliceToString(s pcommon.Slice) string {
+	var b strings.Builder
+	b.WriteByte('[')
+	for i := 0; i < s.Len(); i++ {
+		if i < s.Len()-1 {
+			fmt.Fprintf(&b, "%s, ", attributeValueToString(s.At(i)))
+		} else {
+			b.WriteString(attributeValueToString(s.At(i)))
+		}
+	}
+
+	b.WriteByte(']')
+	return b.String()
+}
+
+func mapToString(m pcommon.Map) string {
+	var b strings.Builder
+	b.WriteString("{\n")
+
+	m.Range(func(k string, v pcommon.Value) bool {
+		fmt.Fprintf(&b, "     -> %s: %s(%s)\n", k, v.Type(), v.AsString())
+		return true
+	})
+	b.WriteByte('}')
+	return b.String()
+}
+
+const (
+	SourceCategoryKey = "_sourceCategory"
+	SourceHostKey     = "_sourceHost"
+	SourceNameKey     = "_sourceName"
+)
+
+func (SumoMarshaler) MarshalLogs(ld plog.Logs) ([]byte, error) {
+	buf := bytes.Buffer{}
+	rls := ld.ResourceLogs()
+	for i := 0; i < rls.Len(); i++ {
+		rl := rls.At(i)
+		sourceCategory, exists := rl.Resource().Attributes().Get(SourceCategoryKey)
+		if !exists {
+			return nil, errors.New("_sourceCategory attribute does not exists")
+		}
+		sourceHost, exists := rl.Resource().Attributes().Get(SourceHostKey)
+		if !exists {
+			return nil, errors.New("_sourceHost attribute does not exists")
+		}
+		sourceName, exists := rl.Resource().Attributes().Get(SourceNameKey)
+		if !exists {
+			return nil, errors.New("_sourceName attribute does not exists")
+		}
+		ills := rl.ScopeLogs()
+		for j := 0; j < ills.Len(); j++ {
+			ils := ills.At(j)
+			logs := ils.LogRecords()
+			for k := 0; k < logs.Len(); k++ {
+				lr := logs.At(k)
+				dateVal := lr.ObservedTimestamp()
+				body := attributeValueToString(lr.Body())
+
+				logEntry(&buf, "{\"date\": \"%s\",\"sourceName\":\"%s\",\"sourceHost\":\"%s\",\"sourceCategory\":\"%s\",\"fields\":{},\"message\":\"%s\"}",
```

Comment:
Are you sure this is going to produce valid JSON? What if I have source field with the value `"foo"bar"` ? Can you please add a test with that example?

I note you're also dropping all resource and log attributes, is that expected or did I miss something?

Reply:
I think it's expected to drop these, but I'll confirm that with others.

Reply:
After verifying the format, I've decided to do the following:
- replace custom field format with JSON and quote it
- place all resource attributes that are not `source*` into `fields`
- merge log record-level attributes into one JSON with the body
Sorry for the mess, but it should be clearer now.